### ISSUE TITLE
Binary wrappers are not JS

### ISSUE THREAD

**Issue by wmertens** [pnpm/pnpm#1282]
### Code to reproduce the issue:

any package that installs a binary, for example `matcha`

### Expected behavior:

`node ./node_modules/.bin/_matcha` works

### Actual behavior:

`./node_modules/.bin/_matcha` is a shellscript and fails.

### Additional information:

Replacing the shell script with javascript like this works fine:

```js
#!/usr/bin/env node
process.env.NODE_PATH = "/Users/wmertens/Documents/AeroFS/Projects/slate/node_modules/.registry.npmjs.org/matcha/0.7.0/node_modules/matcha/node_modules:/Users/wmertens/Documents/AeroFS/Projects/slate/node_modules/.registry.npmjs.org/matcha/0.7.0/node_modules:/Users/wmertens/Documents/AeroFS/Projects/slate/node_modules/.registry.npmjs.org/matcha/node_modules:/Users/wmertens/Documents/AeroFS/Projects/slate/node_modules/.registry.npmjs.org/node_modules:/Users/wmertens/Documents/AeroFS/Projects/slate/node_modules:/Users/wmertens/Documents/AeroFS/Projects/node_modules:/Users/wmertens/Documents/AeroFS/node_modules:/Users/wmertens/Documents/node_modules:/Users/wmertens/node_modules:/Users/node_modules:/node_modules:/Users/wmertens/Documents/AeroFS/Projects/slate/node_modules/matcha/node_modules"
require("../matcha/bin/_matcha")
```

**Comment by wmertens** [comment]
I checked [the docs](https://docs.npmjs.com/files/package.json#bin) and anything can be in these bin scripts, but js is very common.

So I propose that if the name matches `/\.m?js/`, or the first line of the file matches `/^#!.*\/node\b/`, pnpm makes a JS wrapper as above, otherwise it wraps it as before.

The reason is that there are many automated scripts that do things like `node ./node_modules/.bin/foo ...` and the shell-based wrapper breaks that.

**Comment by zkochan** [comment]
Related #736

First of all, even in npm/yarn these scripts are shell scripts on Windows. So this is a more consistent behavior: shell scripts on all systems.

However, consistency is not the reason we have to use shells. We have to override `NODE_PATH` and as far as I remember, your proposed solution did not work. Of course, we can try it. We have tests that cover pluggable tools that need NODE_PATH to be rewritten

**Comment by aparajita** [comment]
@zkochan Can we close this? No activity in two years.

**Comment by zkochan** [comment]
Maybe we should try the suggested solution first.

**Comment by aparajita** [comment]
Well, @zkochan, you have almost 400 open issues going back more than two years, so I'd like to propose establishing some kind of release schedule and a more rigorous triaging process. If you're interested, I'll post it as an issue.

**Comment by zkochan** [comment]
I am open to suggestions. I just think that issues should not be closed just because they've been open for a long time and had no activity.

**Comment by wmertens** [comment]
At my work we use GitHub projects and have a column for wishlist (it's really long); to-do (confirmed things that will happen very soon); in progress; in review. Works nicely

**Comment by bingtimren** [comment]
Was just bitten by this. I have a launch setting that invokes `node --inspect-brk ${workspaceRoot}/node_modules/.bin/jest` to debug test cases. It works under npm / yarn but breaks under pnpm.

**Comment by wmertens** [comment]
@bingtimren I fixed it this way https://github.com/kentcdodds/nps-utils/issues/29 - I do use nps though, it allows more flexibility.

**Comment by rdsedmundo** [comment]
Also facing the same issue under the situation reported by @bingtimren, I can't use `debugger` with Jest.

I work around it by referencing the file directly:
```sh
node --inspect-brk ./node_modules/jest/bin/jest.js --runInBand
```

instead of:
```sh
node --inspect-brk ./node_modules/.bin/jest --runInBand
```

**Comment by woss** [comment]
TBH that doesn't sound that hard or complicated to to. i do it all the time for various other scripts. @rdsedmundo 

from linux point of view what's inside `bin` should be executable and I personally like how pnpm handles the `.bin` related files. you can just call them ( execute them)  without `node` before it.

**Comment by zkochan** [comment]
@rdsedmundo you can set options to Node.js via env variables: https://nodejs.org/api/cli.html#cli_node_options_options

so I guess

```
NODE_OPTIONS="--inspect-brk" pnpx jest --runInBand
```

**Comment by rdsedmundo** [comment]
I'm aware of that env, but unfortunately this doesn't work because it tries to start two debug instances, one for `pnpx` process itself, and other for `jest`:

```
$ NODE_OPTIONS="--inspect-brk" pnpx jest --runInBand

Debugger listening on ws://127.0.0.1:9229/4484c46d-8425-4465-ae64-a00d3cc6b5ba
For help, see: https://nodejs.org/en/docs/inspector
Debugger attached.
Debugger listening on ws://127.0.0.1:49317/e1bc379b-082f-4d24-8826-ada7c6c51973
For help, see: https://nodejs.org/en/docs/inspector
```

:9229 is pnpx's process, :49317 is Jest's.

I agree with @woss, the workaround of just referencing the actual JS file is fine, although you have to see where it's per case basis (packages don't standardize this, whereas the `.bin` path is standard).

**Comment by conartist6** [comment]
`NODE_OPTIONS="--inspect-brk" ./node_modules/.bin/jest` seems like it works fine, and it should work no matter the project given that bin scripts can have different locations and names within a project.

**Comment by conartist6** [comment]
I wonder if the maintainers would be open to making `--inspect` and `--inspect-brk` flags to `pnpx`/`pnpm run` so that the command would just be `pnpm run --inspect jest`?

**Comment by zkochan** [comment]
I guess we can do that. We would just set the NODE_OPTIONS env variable for the script.

**Comment by wmertens** [comment]
I really like @conartist6's solution of setting the NODE_OPTIONS env var; I don't really feel the need for copying Node flags into pnpx (unless npx has them too).

I consider this issue fixed for me now, so unless you want to keep it opent @zkochan, please close.

**Comment by conartist6** [comment]
npx has `npx -n=--inspect [script]`.
yarn has `yarn run --inspect [script]`
ava has `ava debug [test]`

I think the most common operations are worth shorthand, just because this stuff is my job and sometimes I'm doing it with one hand or half asleep and I don't want to have to think about what the magic invocation is to debug a script. I think it is reasonable to expect that my script runner would understand idea that I want to debug the script I am running. I'm willing to write it up as a PR if an acceptable solution can be found.

**Comment by wmertens** [comment]
-n seems like a good addition

**Comment by magicdawn** [comment]
+1 on `pnpm --inspect run`

**Comment by magicdawn** [comment]
NODE_OPTIONS='--inspect-brk' breaks for `pnpm run`

pnpm run use child_process, the pnpm process and the child_process will try to listen the same port 9229.

```
$ NODE_OPTIONS='--inspect-brk' pnpm rollup -c        
Debugger listening on ws://127.0.0.1:9229/4ca4d412-8635-464c-85e9-69dea09f0ef0
For help, see: https://nodejs.org/en/docs/inspector
Debugger attached.
Starting inspector on 127.0.0.1:9229 failed: address already in use
```

example: trying to debug rollup, last line: I guess it's the child_process trying to listen on port 9229

**Comment by conartist6** [comment]
@magicdawn See https://github.com/pnpm/pnpm/issues/1282#issuecomment-1043541144

I'm still very much in favor of `pnpm run --inspect ` and `pnpm run --inspect-brk`

**Comment by magicdawn** [comment]
oh I see, I need to get rid of `pnpm run`

from `NODE_OPTIONS='--inspect-brk' pnpm rollup -c`
to `NODE_OPTIONS='--inspect-brk' ./node_modules/.bin/rollup -c`

**Comment by conartist6** [comment]
Yeah. But it has to be by running a shell script, which is different from the other runners which have you running a node script.

**Comment by conartist6** [comment]
That's one of the major reasons I think it makes so much sense to do `run --inspect` too. That way there can be one thing that we tell developers that will work whatever tool they're using. And making it part of the specified API makes it less likely that there will be unexpected interactions caused by omitting the runner.

**Comment by y-nk** [comment]
because of this, nobody can eventually do `./node_modules/.bin/ts-node ./node_modules/.bin/supposedly-node-script -- some-option-to-enable-ts`

**Comment by y-nk** [comment]
@zkochan thanks a lot for this answer.
